modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java

```diff
@@ -775,4 +799,41 @@ private void renameDeltaFileOnCheckpointThread(GroupPartitionId partitionId) thr
     void updateLastProgressAfterReleaseWriteLock() {
         afterReleaseWriteLockCheckpointProgress = currentCheckpointProgress;
     }
+
+    /**
+     * Callback on destruction of the partition of the corresponding group.
+     *
+     * <p>If the checkpoint is in progress, then wait until it finishes processing the partition that we are going to destroy, in order to
+     * prevent the situation when we want to destroy the partition file along with its delta files, and at this time the checkpoint performs
+     * I/O operations on them.
+     *
+     * @param groupId Group ID.
+     * @param partitionId Partition ID.
+     * @throws IgniteInternalCheckedException If there are errors while processing the callback.
+     */
+    void onPartitionDestruction(int groupId, int partitionId) throws IgniteInternalCheckedException {
+        CheckpointProgressImpl currentCheckpointProgress = this.currentCheckpointProgress;
+
+        if (currentCheckpointProgress == null || !currentCheckpointProgress.inProgress()) {
+            return;
+        }
+
+        CompletableFuture<Void> processedPartitionFuture = currentCheckpointProgress.getProcessedPartitionFuture(groupId, partitionId);
```

Comment:
For a fresh reader, it is not clear why we are not afraid of a race between calling this method and starting a checkpoint. How about adding a comment explaining why this is the case (as you explained me in a private conversation, namely about generation already having been switched, so a new checkpoint would not write any pages from the corresponding group+partition, if I understood correctly).

Reply:
Tried to fix it